core/include/xfa.h

```diff
@@ -0,0 +1,186 @@
+ * It needs to be part of one single compilation unit.
+ *
+ * The pragmas prevent these errors:
+ *
+ *     error: ISO C forbids empty initializer braces
+ *     error: ISO C forbids zero-size array ‘xfatest_end’
+ *
+ * @param[in]   type    name of the cross-file array
+ * @param[in]   name    name of the cross-file array
+ */
+#define XFA_INIT(type, name) \
+    _Pragma("GCC diagnostic push") \
+    _Pragma("GCC diagnostic ignored \"-Wpedantic\"") \
+    _XFA(name, 0_) type name [0] = {}; \
+    _XFA(name, 9_) type name ## _end [0] = {}; \
+    _Pragma("GCC diagnostic pop") \
+    extern const unsigned __xfa_dummy
+
+/**
+ * @brief Declare an external read-only cross-file array
+ *
+ * This macro defines the symbols necessary to use XFA_START() and XFA_END().
+ * Think of this as XFA_INIT() but with "extern" keyword.
+ * It is supposed to be used in compilation units where the cross file array is
+ * being accessed, but not defined using XFA_INIT.
+ *
+ * @param[in]   type    name of the cross-file array
+ * @param[in]   name    name of the cross-file array
+ */
+#define XFA_USE_CONST(type, name) \
+    extern const type name []; \
+    extern const type name ## _end []
+
+/**
+ * @brief Declare an external writable cross-file array
+ *
+ * This macro defines the symbols necessary to use XFA_START() and XFA_END().
+ * Think of this as XFA_INIT() but with "extern" keyword.
+ * It is supposed to be used in compilation units where the cross file array is
+ * being accessed, but not defined using XFA_INIT.
+ *
+ * @param[in]   type    name of the cross-file array
+ * @param[in]   name    name of the cross-file array
+ */
+#define XFA_USE(type, name) \
+    extern type name []; \
+    extern type name ## _end []
+
+/**
+ * @brief Define variable in writable cross-file array
+ *
+ * Variables will end up sorted by prio.
+ *
+ * Add this to the type in a variable definition, e.g.:
+ *
+ *     XFA(driver_params, 0) driver_params_t _onboard = { .pin=42 };
+ *
+ * @param[in]   xfa_name    name of the xfa
+ * @param[in]   prio        priority within the xfa
+ */
+#define XFA(xfa_name, prio) _XFA(xfa_name, 5_ ## prio)
+
+/**
+ * @brief Define variable in read-only cross-file array
+ *
+ * Variables will end up sorted by prio.
+ *
+ * Add this to the type in a variable definition, e.g.:
+ *
+ *     XFA(driver_params, 0) driver_params_t _onboard = { .pin=42 };
+ *
+ * @param[in]   xfa_name    name of the xfa
+ * @param[in]   prio        priority within the xfa
+ */
+#define XFA_CONST(xfa_name, prio) _XFA_CONST(xfa_name, 5_ ## prio)
+
+/**
+ * @brief Add a pointer to cross-file array
+ *
+ * Pointers will end up sorted by prio.
+ *
+ * @param[in]   xfa_name    name of the xfa
+ * @param[in]   prio        priority within the xfa
+ * @param[in]   name        symbol name
+ * @param[in]   entry       pointer variable to add to xfa
+ */
+#define XFA_ADD_PTR(xfa_name, prio, name, entry) \
```

Comment:
There is an uncrustify comment if you want to address it.

Reply:
ran uncrustify over it